server/src/main/java/org/elasticsearch/cluster/coordination/JoinValidationService.java

```diff
@@ -302,14 +302,15 @@ protected void doRun() throws Exception {
             assert discoveryNode.getVersion().onOrAfter(Version.V_8_3_0) : discoveryNode.getVersion();
             // NB these things never run concurrently to each other, or to the cache cleaner (see IMPLEMENTATION NOTES above) so it is safe
             // to do these (non-atomic) things to the (unsynchronized) statesByVersion map.
-            final var cachedBytes = statesByVersion.get(discoveryNode.getVersion().transportVersion);
+            TransportVersion version = transportService.getConnection(discoveryNode).getTransportVersion();
+            final var cachedBytes = statesByVersion.get(version);
             final var bytes = Objects.requireNonNullElseGet(cachedBytes, () -> serializeClusterState(discoveryNode));
             assert bytes.hasReferences() : "already closed";
             bytes.incRef();
             transportService.sendRequest(
```

Comment:
I'd prefer to use the `sendRequest` which takes the connection to use here, given that we have already looked up the connection by this point.